#Python Exercise 3
#Samuel Statton

In [1]:
from google.colab import drive 
drive.mount('/content/drive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&response_type=code&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly

Enter your authorization code:
··········
Mounted at /content/drive


In [0]:
import pandas as pd
import sklearn
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LassoLarsCV
import matplotlib.pyplot as plt

In [0]:
alldata = pd.read_csv('drive/My Drive/Colab Notebooks/finalmaster-ratios.csv')

In [4]:
alldata.head()

# Purchases  B01001001   B01001002  ...  B19001015  B19001016  B19001017
0           22     206252  469.226965  ...  36.440765  23.446284  21.197485
1            7      61399  486.538869  ...  33.000508  33.169741  24.792689
2            3      73170  489.859232  ...  20.125056  11.890525  16.537397
3           94     251724  505.585483  ...  43.731067  38.851729  40.427349
4            0      37382  495.586111  ...  36.118530  31.603714  19.648989

[5 rows x 190 columns]

In [0]:
allvariablenames = list(alldata.columns.values)

#removing first 8 (0-7) variables
listofallpredictors = allvariablenames[8:-1]

In [0]:
#load predictors into dataframe
predictors = alldata[listofallpredictors]  

#load target into dataframe
target = alldata['# Purchases']  

In [0]:
pred_train, pred_test, tar_train, tar_test = train_test_split(predictors, target, test_size=.3, random_state=123) 

In [55]:
model = sklearn.linear_model.LassoLarsCV(cv=10, precompute=False).fit(pred_train, tar_train)

/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/least_angle.py:578: ConvergenceWarning: Regressors in active set degenerate. Dropping a regressor, after 13 iterations, i.e. alpha=6.800e-01, with an active set of 11 regressors, and the smallest cholesky pivot element being 6.322e-08. Reduce max_iter or increase eps parameters.
  ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/least_angle.py:578: ConvergenceWarning: Regressors in active set degenerate. Dropping a regressor, after 25 iterations, i.e. alpha=3.400e-01, with an active set of 23 regressors, and the smallest cholesky pivot element being 6.322e-08. Reduce max_iter or increase eps parameters.
  ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/least_angle.py:578: ConvergenceWarning: Regressors in active set degenerate. Dropping a regressor, after 47 iterations, i.e. alpha=1.839e-01, with an active set of 37 regressors, and the smallest cholesky pivot eleme

##Question 1:

In [56]:
#creating a data frame based on the updated list of predictors
predictors_model=pd.DataFrame(listofallpredictors)

#assigning the name 'label' to the predictors
predictors_model.columns = ['label']

#creating a 'coeff' column that is assigned the coefficient values from the LassoLarsCV model 
predictors_model['coeff'] = model.coef_

#this for loop and if statement is going row by row (or predictor by predictor)
#and selecting all coefficients greater than zero and printing both the label and coefficient
for index, row in predictors_model.iterrows():
    if row['coeff'] > 0:
        print(row.values)
      

['B01001014' 1.3142095900947361]
['B01001036' 3.199831473110331]
['B01001037' 0.5867114750885728]
['B01001038' 1.0295045766028337]
['B02001005' 1.2648999446336935]
['B13014026' 0.8858208978839138]
['B13014027' 1.2507563632390104]
['B13016008' 0.48413572331173443]
['B15002015' 0.5457730397237327]
['B15002027' 2.01281915236849]
['B19001016' 0.04656593306654476]


##Question 1 Continued:

Predictors and their census data:

['B01001014' 1.3142095900947361] -- Males 40-44 years old

['B01001036' 3.199831473110331] -- Females 30-34 years old

['B01001037' 0.5867114750885728] -- Females 35-39 years old

['B01001038' 1.0295045766028337] -- Females 40-44 years old

['B02001005' 1.2648999446336935] -- Race: Asian Alone

['B13014026' 0.8858208978839138] -- Women 15-50 who have not given birth in the past 12 months, unmarried, and have a Bachelor's Degree


['B13014027' 1.2507563632390104] -- Women 15-50 who have not given birth in the past 12 months, unmarried, and have a Graduate or Profession Degree


['B13016008' 0.48413572331173443] -- Women 40-44 who have given birth in the past 12 months

['B15002015' 0.5457730397237327] -- Males over 25 with a Bachelor's Degree

['B15002027' 2.01281915236849] -- Females 25 or over with no diploma/12th grade highest attainment

['B19001016' 0.04656593306654476] -- Household with income 150,000-199,999 in last 12 months

##Question 2:


###Age & Gender Only Predictors
Females 30-34 (3.20) -- Meaning that we sell more Bobo Bars in areas with females 30-34 (Example you gave) but also could be the leading age and gender demographic for audience segmentation. We are also popular with males 40-44 (1.31) and are followed by females 40-44 (1.03) which indicates we sell more well in markets with high numbers of people in their early 40's. The one segment we struggle with is females 35-39 (0.587). 

###Race
We do well in markets with Asian populations (1.26)

###Nuanced Age and Gender
When looking at females that are unmarried and have not given birth in the past 12 months, we do significantly better with women that have a Graduate/Profession Degree compared to only a Bachelor Degree (1.25 v. 0.886). We also do poorly with females 40-44 who have given birth (0.484) which is a stark difference when we just look at females 40-44 (1.03). Females over 25 with no diploma (2.01) We also sell more Bobo Bars to females over 25 with basic education while we struggle with males over 25 with a bachelors (0.546).

###Income
We do poorly with household incomes between 150k and 200k. 



##Question 3:

I would give my boss females 30-34 (3.20) and females with no diploma (2.01)

In [57]:
train_error = sklearn.metrics.mean_squared_error(tar_train, model.predict(pred_train))
print ('training data MSE')
print(train_error)

training data MSE
20798.061542418778


In [58]:
train_error = sklearn.metrics.mean_squared_error(tar_test, model.predict(pred_test))
print ('test data MSE')
print(train_error)

test data MSE
39002.73812326359


##Question 4:
The test data has a significantly higher mean squared error which means the model is not performing well with the test data. This means that the model is overfitting and causing the error to increase on the test data. 

In [59]:
#r squared
rsquared_train=model.score(pred_train,tar_train)
print ('training data R-square')
print(rsquared_train)

training data R-square
0.2823739261902274


In [60]:
#r squared
rsquared_train=model.score(pred_test,tar_test)
print ('test data R-square')
print(rsquared_train)

test data R-square
0.22637894007738157


##Question 5:
The model only explains about 22.5% to 28% of our sales which is not very accurate. It may give insights into our most profitable customer segments, but it cannot reliably predict sales numbers solely on census data. 

In [61]:
print("y interecept:")
print(model.intercept_)
listofallpredictors[0]

y interecept:
130.29546325145463


'B01001008'

##Question 6:
The baseline sales is the number of purchases with the omited variable of males 20 years old. So when we read the coefficients for questions 1 & 2, the increase or decrease in sales is relative to the 20 year old male segment. 